Environment detection

gbirke opened the pull request:
https://phabricator.wikimedia.org/T351627

timEulitz commented on desktop/banner_var.js:
```diff
@@ -263,6 +315,7 @@ $( '#bImpCount' ).val( bannerImpCount );
 // Display banner on load
 $( function () {
 	var $bannerElement = $( '#WMDE_Banner' );
+	var $minibannerElement = $( '#mini-anner' );
```
Should probably be `#mini-banner`.

moiikana replied:
thanks for finding that

gbirke committed:
Centralize Runtime Environment detection

We want to push the decision if we're in dev mode to the entry points
and want to check only once. This is a preparatory refactoring that'll
allow us to determine the "remainingImpressions" based on the runtime
environment.

Ticket: https://phabricator.wikimedia.org/T351627

gbirke committed:
Move remainingImpressions calculation

This is now part of the ImpressionCount implementations and
LocalImpressionCount checks the runtime environment and always returns
the max count when in dev mode.

Ticket: https://phabricator.wikimedia.org/T351627

gbirke commented:
The new dev-mode-aware getImpressions is currently only used in the Desktop. I'd like to wait for approval/change requests before I change all the other banners.

Abban commented:
LGTM

moiikana commented on src/utils/RuntimeEnvironment.ts:
```diff
@@ -0,0 +1,54 @@
+}
+
+export class UrlRuntimeEnvironment implements RuntimeEnvironment {
+	private _isInDevMode: boolean;
+	private _runsInDevEnvironment: boolean;
```
what are the different purposes of these two fields?

Abban replied:
One marks the `?devMode` query for previewing the banners on WPORG, the other is the `?devBanner` query item from the Fun Forge
